# 4.5 Data Consistency Checks

In [236]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Importing Dataframes

In [237]:
path = r'c:\Users\graha\Desktop\Career foundry\Chapter 4\Instacart Basket Analysis 09.02.2024'

In [238]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [239]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Excercise 4.5

In [240]:
# create a Dataframe
df_test = pd.DataFrame()

In [241]:
#create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [242]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [243]:
#checking for mixed-type columns in the dataframe
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [244]:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [245]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [246]:
#Creating a Dataframe with the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [247]:
#printing the data frame with missing values
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14      12
68             69          NaN        26              7      12
115           116          NaN        93              3      11
261           262          NaN       110             13      12
525           525          NaN       109             11       1
1511         1511          NaN        84             16      14
1780         1780          NaN       126             11      12
2240         2240          NaN        52              1      14
2586         2586          NaN       104             13      12
3159         3159          NaN       126             11      13
3230         3230          NaN       120             16      14
3736         3736          NaN        41              8      15
4283         4283          NaN        77              7      14
4790         4790          NaN        91             16      14
38187       38183          NaN        39             12      21
40444       40440          NaN       120             16      15

In [248]:
#finding the mean 
df.describe() 

Unnamed: 0  order_id  user_id  order_number  order_dow  \
count     3421083   3421083  3421083       3421083    3421083   
mean      1710541   1710542   102978            17          3   
std        987582    987582    59534            18          2   
min             0         1        1             1          0   
25%        855270    855272    51394             5          1   
50%       1710541   1710542   102689            11          3   
75%       2565812   2565812   154385            23          5   
max       3421082   3421083   206209           100          6   

       order_hour_of_day  days_since_prior_order  
count            3421083                 3214874  
mean                  13                      11  
std                    4                       9  
min                    0                       0  
25%                   10                       4  
50%                   13                       7  
75%                   16                      15  
max                   23                      30

In [249]:
df_prods.shape

(49693, 5)

In [250]:
#cleaning the Dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [251]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [252]:
#creating the dataframe of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [253]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11       5  
18459        27              5       9  
26810        27              5      13  
35309       121             14       7  
35495       127             11      10

In [254]:
#Adressing duplicates
df_prods_clean.shape

(49677, 5)

In [255]:
#removing the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [256]:
df_prods_clean_no_dups.shape

(49672, 5)

In [257]:
#exporting the cleaned data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Excecise 4.5

# step 2

In [258]:

#Describing the initial Dataframe
df_ords.describe()

Unnamed: 0  order_id  user_id  order_number  order_dow  \
count     3421083   3421083  3421083       3421083    3421083   
mean      1710541   1710542   102978            17          3   
std        987582    987582    59534            18          2   
min             0         1        1             1          0   
25%        855270    855272    51394             5          1   
50%       1710541   1710542   102689            11          3   
75%       2565812   2565812   154385            23          5   
max       3421082   3421083   206209           100          6   

       order_hour_of_day  days_since_prior_order  
count            3421083                 3214874  
mean                  13                      11  
std                    4                       9  
min                    0                       0  
25%                   10                       4  
50%                   13                       7  
75%                   16                      15  
max                   23                      30

In [259]:
#changing the format of the results out of scientific notation so they are easy to read
pd.options.display.float_format = '{:.0f}'.format

In [260]:
#Success
df_ords.describe()

Unnamed: 0  order_id  user_id  order_number  order_dow  \
count     3421083   3421083  3421083       3421083    3421083   
mean      1710541   1710542   102978            17          3   
std        987582    987582    59534            18          2   
min             0         1        1             1          0   
25%        855270    855272    51394             5          1   
50%       1710541   1710542   102689            11          3   
75%       2565812   2565812   154385            23          5   
max       3421082   3421083   206209           100          6   

       order_hour_of_day  days_since_prior_order  
count            3421083                 3214874  
mean                  13                      11  
std                    4                       9  
min                    0                       0  
25%                   10                       4  
50%                   13                       7  
75%                   16                      15  
max                   23                      30

# step 3

In [261]:
#Check for Mixed data types - "applymap" was giving me errors so i used "map" Instead
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print ((col))

In [262]:
#checking to see if there are any duplicates
weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].map(type)).any(axis=1)
print(weird)

0    False
1    False
2    False
3    False
dtype: bool


# step 4

In [263]:
#there are no duplicates

# Step 5

In [264]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### The days_since_prior_order has 206209 missing values this is likely because either people havent placed an order or they have placed an order on that day which could potentially return Null values

# Step 6

In [265]:
# we need to clean the data, one way could be to change the data into a string,
# however this is not a string value so that wouldnt be optimal

In [266]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [267]:
df_ords.shape

(3421083, 8)

In [268]:
# addressing the null values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [269]:
#success
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

In [270]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
1                 1   2398795        1    prior             2          3   
2                 2    473747        1    prior             3          3   
3                 3   2254736        1    prior             4          4   
4                 4    431534        1    prior             5          4   
...             ...       ...      ...      ...           ...        ...   
3421078     3421078   2266710   206209    prior            10          5   
3421079     3421079   1854736   206209    prior            11          4   
3421080     3421080    626363   206209    prior            12          1   
3421081     3421081   2977660   206209    prior            13          1   
3421082     3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                      15  
2                       12                      21  
3                        7                      29  
4                       15                      28  
...                    ...                     ...  
3421078                 18                      29  
3421079                 10                      30  
3421080                 12                      18  
3421081                 12                       7  
3421082                 14                      30  

[3421083 rows x 8 columns]

### I used this method becuase i want to view the data in one Datafrome. There are far too many entries to remove from the data set. There are also perfectly reasonable explanations as to why this column has null values, such as clients who place one order or orders that were placed on the day this data was extracted. thus i will not remove or alter the data

# Step 7

In [271]:
#checking for duplicates
df_ords_dup = df_ords[df_ords.duplicated()]

In [272]:
#Checking the results
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

### we can see from the results there are no duplicates no action is required

In [278]:
df_ords = df_ords.drop(columns = ['eval_set'])

KeyError: "['eval_set'] not found in axis"

In [ ]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [ ]:
df_ords.head()

In [279]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords4.5.csv'))

In [280]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods4.5.csv'))